stack-graphs/src/storage.rs

```diff
@@ -0,0 +1,472 @@
+        tag: &str,
+    ) -> Result<()> {
+        let file_str = graph[file].name();
+        copious_debugging!("--> Add graph for {}", file_str);
+        let graph = serde::StackGraph::from_graph_filter(graph, &FileFilter(file));
+        self.conn.execute("BEGIN;", ())?;
+        // insert or update graph
+        let mut stmt = self
+            .conn
+            .prepare_cached("INSERT OR REPLACE INTO graphs (file, tag, json) VALUES (?, ?, ?)")?;
+        stmt.execute((file_str, tag, &serde_json::to_vec(&graph)?))?;
+        // remove stale file paths
+        let mut stmt = self
+            .conn
+            .prepare_cached("DELETE FROM file_paths WHERE file = ?")?;
+        stmt.execute([file_str])?;
+        // remove stale file paths
+        let mut stmt = self
+            .conn
+            .prepare_cached("DELETE FROM root_paths WHERE file = ?")?;
+        stmt.execute([file_str])?;
+        self.conn.execute("END;", ())?;
+        Ok(())
+    }
+
+    /// Add a partial path for a file to the database.  Throws an error if the file does not exist in
+    /// the database.
+    pub fn add_partial_path_for_file(
+        &mut self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        path: &PartialPath,
+        file: Handle<File>,
+    ) -> Result<()> {
+        let file_str = graph[file].name();
+        copious_debugging!(
+            "--> Add {} partial path {}",
+            file_str,
+            path.display(graph, partials)
+        );
+        let start_node = graph[path.start_node].id();
+        if start_node.is_in_file(file) {
+            copious_debugging!(
+                " * Add as node path from node {}",
+                path.start_node.display(graph),
+            );
+            let local_id = start_node.local_id();
+            let path = serde::PartialPath::from_partial_path(graph, partials, path);
+            let mut stmt = self
+                .conn
+                .prepare_cached("INSERT INTO file_paths (file, local_id, json) VALUES (?, ?, ?)")?;
+            stmt.execute((file_str, local_id, &serde_json::to_vec(&path)?))?;
+        } else if start_node.is_root() {
+            copious_debugging!(
+                " * Add as root path with symbol stack {}",
+                path.symbol_stack_precondition.display(graph, partials),
+            );
+            let symbol_stack = path.symbol_stack_precondition.storage_key(graph, partials);
+            let path = serde::PartialPath::from_partial_path(graph, partials, path);
+            let mut stmt = self.conn.prepare_cached(
+                "INSERT INTO root_paths (file, symbol_stack, json) VALUES (?, ?, ?)",
+            )?;
+            stmt.execute((file_str, symbol_stack, &serde_json::to_vec(&path)?))?;
+        } else {
+            panic!(
+                "added path {} must start in given file {} or at root",
+                path.display(graph, partials),
+                graph[file].name()
+            );
```

Comment:
General question: when a function can `panic`, should that be indicated in the comment for the function? I'm not sure if there is a Rust convention for that, and this isn't intended to be a suggestion here, it's only a drive-by question.

Reply:
Good question. I think you're right that it is a good idea to add that in the comment!